### ISSUE TITLE
Add operation for opening a URL with given command

### ISSUE THREAD

**Issue by Minoru** [newsboat/newsboat#1171]
As @der-lyse explains in https://github.com/newsboat/newsboat/issues/1165#issuecomment-685939764:

> One thing which might break user expectations is the abort on the first failed command in a sequence. When using cleanups in the last step this would cause them to not be executed and leaving newsboat in a little bit messed up state. Suppose a [common and documented use case in macros](https://newsboat.org/releases/2.20.1/docs/newsboat.html#_macro_support) where people change their `browser` setting to a different executable, opening the article or link in the new "browser" and finally reverting back to their default browser. In case `open-in-browser` fails, the `browser` is not reset, which is probably not what one would expect:
>
>    macro i set browser "feh %u"; open-in-browser ; set browser "elinks %u"
>
> However, if I followed the development correctly, macros were changed lately to behave the same, so this potential problem might not be that severe. The temporary browser change seems like a total hack to me, anyways – so we should probably add proper support for the use case of running a specific command on a given URL (but this is off-topic).

We can replace the whole thing with a new `open-with-command`, to be used like this: `open-with-command "feh %u"` (the name is subject to bikeshedding).

**Comment by m040601** [comment]
Just want to chime in, as a long time (power) newsboat end user, to upvote this idea/proposal, and give some feedback. Sorry I'm not a developer, can't comment or provide any coding ideas/directions.

> The temporary browser change seems like a total hack to me

It's been, like, years, that I (and many others) have felt the same.
Many times I had thought about submiting a question/issue about this.But somehow always thought that this was something "intrinsic" to the way newsboat was designed from the first day. And that therefore it would be difficult or impossible to change it

There are always many things to improve. And of course one has to choose what to prioritize for development. From an end user point of view I would say that since,

``macro X set browser "ABC"; open-in-browser ; set browser "back to DEF"
`

is one of the most used, and hacked config lines to "bend" newsboat to do what you want.

Anyway, this is just to thank you all, for the amazing work that you've been doing lately and the carefull thought given to this kind

`

**Comment by A1RO** [comment]
Hi @m040601, thanks for the thorough explanation!

To add on your idea:

> I even think that many people new to newsboat, fail to grasp the power of that config line, because of the words "browser" and "open" there. It reads as just "see your article displayed on another web browser".

Use cases like the one I described in #1175, i.e. `macro p set browser "mpv"; one ; set browser "firefox"` certainly does not read as "open this podcast in your favorite player". A little cryptic there, and it's only a simplified version of my actual command. I mean, it's fun creating it, but also represents a steep learning curve for the less technically inclined users. The fact that `mpv` is a _media player_ (not a _browser_) is already confusing enough. Same goes for `feh` which is an _image viewer_.

I would say that's a good enough reason to keep a generic name for the new operation.

Another point i'd like to cover is the ability to choose **which link** (or links) to pass to the external command. For example, something like `open-with "mpv %1"` would fit the bill perfectly. Another way could be `open-with mpv 1`. Not specifying the URL would use the actual implementation of the _browser_ optional `%u`.

Being able to open any link from the article list would also be a timesaver, but not a priority.

On this topic:

> But if one reads it as "operate" (what to do) on the url ( the input) to get some other stuff (the output) then the all the power of Unix scripting and piping possibilities becomes more evident. I myself, remember, that only after seeing many different people configs with cool, clever and creative uses of set browser "do stuffz ABC" did I realize that.

Documenting a few more examples of newsboat typical usage would solve that.

Off the top of my head, writing about those use cases in the main documentation could help many users while spreading the word about newsboat flexibility:

- How to open links in a _text-based web browser_
- How to open links in a _graphical web browser_
- How to open images in an _image viewer_
- How to open audio and video files in a _media player_
- How to _download_ linked files to disk 
- How to launch command in a separate _terminal_
- How to _isolate_ command and keep on using newsboat while it's running

I could propose a draft about those items if that's the intended direction of the project.
It might also be wise to wait and see for the new operation functionalities before suggesting even more hacks.

**Comment by Minoru** [comment]
@m040601, thanks for another perspective on this! The name for the new operation certainly needs some more thought put into it, to make it generic yet useful enough.

@A1RO, good that you mentioned #1175, I forgot about that use-case already ._. I actually like your `mpv %1` idea, but here's yet another: the new command could be something like `with <setting name> <setting value> <operation>`, to be used like this:

```
macro a with browser "feh %u" open
macro p with browser "mpv" one
```

That new operation would work just like `with` in Python, or `bracket` in Haskell: it remembers current value of the setting, changes it to the specified value, runs the op, and restores the original value of the setting (even if the operation failed). But I fear this might be a bit too generic and hard to understand for non-programmers.

> I can propose a draft about those items if that's the intended direction of the project.

I'm not sure adding more examples really helps, but perhaps it's just my lack of imagination :) Please add a couple and submit a PR for me to see what you mean.

> How to isolate command and keep on using newsboat while it's running

We have an FAQ item on this, by the way: https://newsboat.org/releases/2.20.1/docs/faq.html#_after_i_open_a_link_in_the_browser_newsboat_stays_unresponsive_until_the_browser_is_closed

> @m040601: There are always many things to improve. And of course one has to choose what to prioritize for development.
> @A1RO: But it might also be wise to wait and see for the new operation before suggesting more hacks...

Just to make sure you don't have unrealistic expectations here: all the people who work on Newsboat are volunteers, so I don't have power to *make* them do something. All I can do is guide the efforts and reject the ones I don't agree with. My own contributions are usually about upkeep, not introducing new features. As a consequence, we don't do priorities; there's no way to force anyone to heed them anyway. For the same reason, we don't have timelines, ETAs, mostly don't tie features to milestones, etc.

Instead, we just follow the interests of individual developers. If someone is interested in working on this, it will be done. Otherwise it won't be.

I understand this puts non-programmers at a hu-u-uge disadvantage (all they can do is wait), but that's just how things are.

**Comment by m040601** [comment]
This is really an interesting discussion.
And one of those, which goal might not be get an "immediate answer" but is very usefull in thinking about the real power and functionality of newsboat.

I'm gonna leave some more additional thoughts, comments and examples here, so that it can be referenced and easly found in the future.


Thinking deeper about the bookmarking functionality, the boomark script, and what that really entails.

Just like I argued before for the "open in browser" concept

..

> .. because of the words "browser" and "open" there....but if one reads it as "operate" (what to do) on the url ( the input) to get some other stuff (the output) then the all the power of Unix scripting and piping possibilities becomes more evident.


I just realized that with the "bookmarking" concept, the same thing happens. That is one sees the word "bookmark", and might forget that this can be in fact a much more powerfull operation. That is, 

take that "data set"
TITLE
URL
DESC
COMMENT
and "operate" on that

So again, endless Unix pipelining possibilities and destinations.


I got this thought when I was revisiting this guy's Newsboat "bookmarking" setup:

> uriel1998/newsbeuter-dangerzone: Enhanced, modular, bookmarking for newsboat, newsbeuter, or (for that matter) anything that can pass a title and an URL to a program - even from your clipboard.


https://github.com/uriel1998/newsbeuter-dangerzone

[oysttyer/oysttyer: An interactive console text-based command-line Twitter client written in Perl](https://github.com/oysttyer/oysttyer)
